lib/tasks/support/molecule_structure_curation.rb

```diff
@@ -0,0 +1,260 @@
+#   for each molecule
+#     -> find associated samples:                                                             get_samples
+#        -> if sample has molfile, try to recreate the Molecule                               get_or_create_molecule
+#           -> if molecule is found, update the Sample
+#
+#     -> reload associated samples                                                            get_samples
+#       -> if samples left, try to create a new molecule using the molecule cano_smiles       get_new_molecule
+#          -> if new molecule is found, update the remaining samples
+#          -> if new molecule is not found
+#              -> TODO: set remaining samples as decoupled and set the mw
+#
+#     -> if no samples left, destroy the molecule
+#
+
+class MoleculeStructureCuration
+  def process
+    faulty_molecules.find_each do |molecule|
+      next if samples_empty?(molecule)
+
+      molecule_id = molecule.id
+      # try to recreate a molecule for each sample from sample.molfile
+      process_samples_by_sample_mol(molecule)
+      next if samples_empty?(molecule_id)
+
+      # try to create a new molecule using the molecule cano_smiles
+      # then update the remaining samples
+      process_samples_by_molecule_smiles(molecule)
+      next if samples_empty?(molecule_id)
+
+      # convert samples to decoupled samples
+      # set the mw
+      #
+      # destroy the molecule
+      logger.info "Failed to create new molecule for molecule with id: #{molecule_id}"
+    end
+  end
+
+  # try to recreate a molecule for each sample from sample.molfile for all samples
+  # associated with the given molecule
+  # @param [Molecule]
+  # @return [void]
+  def process_samples_by_sample_mol(molecule)
+    logger.info __method__
+    get_samples(molecule).each do |sample|
+      logger.info "Processing sample with id: #{sample.id}"
+      molfile = sample.molfile
+      next if decoupled?(sample)
+      next if faulty_molfile?(molfile)
+      next if get_inchikey(molfile).blank?
+
+      new_molecule = get_or_create_molecule(molfile)
+      if new_molecule.blank? || new_molecule.id == sample.molecule_id
+        logger.info "Molecule not found or already associated with sample with id: #{sample.id}"
+        next
+      end
+      update_sample(sample, new_molecule)
+    end
+  end
+
+  # try to create a new molecule using the molecule cano_smiles
+  # the remaining samples will be updated with the new molecule
+  # @param [Molecule]
+  # @return [void]
+  def process_samples_by_molecule_smiles(molecule)
+    molecule_id = molecule.id
+    logger.info __method__
+    new_molecule = get_new_molecule(molecule)
+    return if new_molecule.blank?
+
+    get_samples(molecule_id).each { |sample| update_sample(sample, new_molecule) }
+  end
+
+  # update the given sample with the new Molecule
+  # @param [Sample]
+  # @param [Molecule]
+  # @return [void]
+  def update_sample(sample, molecule)
```

Comment:
What if the newly created molecule has a different molecular weight and other different attributes (specifically: molecular_mass in Sample model which is mirrored from molecular_weight check Sample.js line 902)?, is there an update needed? or shall we add this to the log info?

Reply:
I will add it to the log